### ISSUE TITLE
Jellyfin does not retry IPv4 on IPv6 failure

### ISSUE THREAD

**Issue by brianjmurrell** [jellyfin/jellyfin#1002]
Getting this trying to view the Plugin Catalog on 10.2.1:
```
[2019-02-25 05:25:48.807 -05:00] [INF] "HTTP GET" "https://videoserver:8920/emby/Packages?TargetSystems=Server&IsAdult=false&IsAppStoreSafe=true&PackageType=UserInstalled". UserAgent: "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
[2019-02-25 05:25:48.825 -05:00] [INF] HttpClientManager "GET": "https://repo.jellyfin.org/releases/plugin/manifest.json"
[2019-02-25 05:25:48.827 -05:00] [INF] "HTTP GET" "https://videoserver:8920/emby/Plugins". UserAgent: "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.119 Safari/537.36"
[2019-02-25 05:25:48.828 -05:00] [INF] HTTP Response 200 to "fd31:aeb1:48df:0:3b14:e643:83d8:7017". Time: "1""ms". "https://videoserver:8920/emby/Plugins" ""
[2019-02-25 05:26:08.828 -05:00] [ERR] Error Timeout getting response from "https://repo.jellyfin.org/releases/plugin/manifest.json"
System.Net.WebException: The operation has timed out.
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
[2019-02-25 05:26:08.853 -05:00] [ERR] Error processing request
MediaBrowser.Model.Net.HttpException: The operation has timed out. ---> System.Net.WebException: The operation has timed out.
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
   --- End of inner exception stack trace ---
   at Emby.Server.Implementations.HttpClientManager.HttpClientManager.SendAsyncInternal(HttpRequestOptions options, String httpMethod) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/HttpClientManager/HttpClientManager.cs:line 403
   at Emby.Server.Implementations.HttpClientManager.HttpClientManager.SendAsync(HttpRequestOptions options, String httpMethod) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/HttpClientManager/HttpClientManager.cs:line 241
   at Emby.Server.Implementations.Updates.InstallationManager.GetAvailablePackagesWithoutRegistrationInfo(CancellationToken cancellationToken) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/Updates/InstallationManager.cs:line 186
   at Emby.Server.Implementations.Updates.InstallationManager.GetAvailablePackages(CancellationToken cancellationToken, Boolean withRegistration, String packageType, Version applicationVersion) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/Updates/InstallationManager.cs:line 175
   at MediaBrowser.Api.PackageService.Get(GetPackages request) in /builddir/build/BUILD/jellyfin-10.2.1/MediaBrowser.Api/PackageService.cs:line 157
   at Emby.Server.Implementations.Services.ServiceExecGeneral.GetTaskResult(Task task) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/Services/ServiceExec.cs:line 110
   at Emby.Server.Implementations.Services.ServiceHandler.ProcessRequestAsync(HttpListenerHost appHost, IRequest httpReq, IResponse httpRes, ILogger logger, String operationName, CancellationToken cancellationToken) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/Services/ServiceHandler.cs:line 95
   at Emby.Server.Implementations.HttpServer.HttpListenerHost.RequestHandler(IHttpRequest httpReq, String urlString, String host, String localPath, CancellationToken cancellationToken) in /builddir/build/BUILD/jellyfin-10.2.1/Emby.Server.Implementations/HttpServer/HttpListenerHost.cs:line 608
[2019-02-25 05:26:08.874 -05:00] [INF] HTTP Response 500 to "fd31:aeb1:48df:0:3b14:e643:83d8:7017". Time: "20066""ms (slow)". "https://videoserver:8920/emby/Packages?TargetSystems=Server&IsAdult=false&IsAppStoreSafe=true&PackageType=UserInstalled" ""
```

But it fetches just fine from here:

```
$ curl -q -s -v -o /dev/null 'https://repo.jellyfin.org/releases/plugin/manifest.json'
*   Trying 2602:ffb6:2:0:f816:3eff:fef9:65ca...
* TCP_NODELAY set
*   Trying 172.81.180.233...
* TCP_NODELAY set
* Connected to repo.jellyfin.org (172.81.180.233) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [108 bytes data]
* TLSv1.2 (IN), TLS handshake, Certificate (11):
{ [2562 bytes data]
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
{ [300 bytes data]
* TLSv1.2 (IN), TLS handshake, Server finished (14):
{ [4 bytes data]
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
} [37 bytes data]
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.2 (OUT), TLS handshake, Finished (20):
} [16 bytes data]
* TLSv1.2 (IN), TLS handshake, Finished (20):
{ [16 bytes data]
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=repo.jellyfin.org
*  start date: Jan 25 23:06:40 2019 GMT
*  expire date: Apr 25 23:06:40 2019 GMT
*  subjectAltName: host "repo.jellyfin.org" matched cert's "repo.jellyfin.org"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
} [5 bytes data]
> GET /releases/plugin/manifest.json HTTP/1.1
> Host: repo.jellyfin.org
> User-Agent: curl/7.61.1
> Accept: */*
> 
{ [5 bytes data]
< HTTP/1.1 200 OK
< Server: nginx/1.13.10
< Date: Mon, 25 Feb 2019 10:29:58 GMT
< Content-Type: application/json
< Content-Length: 11913
< Last-Modified: Mon, 18 Feb 2019 20:29:36 GMT
< Connection: keep-alive
< ETag: "5c6b15b0-2e89"
< Strict-Transport-Security: max-age=31536000
< Accept-Ranges: bytes
< 
{ [11913 bytes data]
* Connection #0 to host repo.jellyfin.org left intact
```

Notice however that when using `curl`, the IPv6 connection fails and it does the right thing and tries other addresses returned for the name in the DNS query.  Jellyfin is NOT doing this.  It's giving up after trying only the first address returned.

Jellyfin (and any IP client in fact) should be doing the same thing and trying all of the addresses return from a gethost*() not just the first one.

**Comment by jomp16** [comment]
As a workaround, put `172.81.180.233 repo.jellyfin.org` on /etc/hosts in Docker container.

**Comment by anthonylavado** [comment]
@joshuaboniface Can you see if this still applies?

**Comment by anthonylavado** [comment]
We removed the AAAA record because there were too many people affected by the Cogent/HE split.

**Comment by brianjmurrell** [comment]
Removing the AAAA record is a work-around **at best**.  It doesn't do anything about the fact that jellyfin is a bad TCP/IP client for not retrying on all available IP addresses.

I would ask that you re-open this ticket because that's what this issue it's about.  It's not about the symptoms which is the failure to get the item, etc.

**Comment by joshuaboniface** [comment]
I agree @brianjmurrell - I've updated the issue description to match.

**Comment by brianjmurrell** [comment]
@joshuaboniface While I agree that IPv4 fallback after IPv6 failure is one of the possible _try all of the addresses_ scenarios that Jellyfin is not doing, I suspect it's even more generic than that.

I'd hazard to bet that if multiple IPv4 addresses are returned from a name lookup and connecting to the first one failed, the remaining addresses would not be tried.  So same problem, just different circumstances and doesn't involve IPv6 at all.

Not that I particularly care about the subject of the ticket (other than perhaps for search results by the next person to discover this) I just don't want the person who picks this up to be fooled into thinking this is an IPv6-only issue.

Or ticket triage mistakenly dropping this in a _low priority because it only affects IPv6 use cases_ bucket.

But all of that aside, much thanks for your open-mindedness and reopening this ticket.

**Comment by oddstr13** [comment]
We would probably want something like [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs) implemented.

About the "low priority" on triage, we only really had *high* and *low*, *high* being for critical bugs that should've been patched yesterday, the only other choice being *low*.

**Comment by brianjmurrell** [comment]
@oddstr13 Happy eyeballs certainly is one way.  Probably the best these days.  But even just a traditional old fallback would be nicer than what's there now.

Fair enough about the priorities.  Unfortunately it's all too often that issues that are seen to only affect IPv6 are downplayed as they only affect "a small percentage" of people when in fact IPv6 [adoption is growing](https://www.google.com/intl/en/ipv6/statistics.html#tab=ipv6-adoption).

**Comment by oddstr13** [comment]
>Fair enough about the priorities. Unfortunately it's all too often that issues that are seen to only affect IPv6 are downplayed as they only affect "a small percentage" of people when in fact IPv6 adoption is growing.

Personally I've been using IPv6 the past 10 years, natively the past 4-5 years.

> In any case, I just didn't want this issue to be confused as applying to IPv6 only because it (likely) doesn't. It would apply to IPv4 -> IPv4 fallback also.

Most likely

**Comment by joshuaboniface** [comment]
And I help deploy it to ISPs. Don't worry you're in a good crowd for IPv6 support ;-)

That said, like every issue in Jellyfin, it required someone to dig into it and provide a fix - the prioritization is more for "how much should we pester people to find someone to work on it" than it is for "the team says this is low priority". We're hoping to reserve that "high" priority for major breaking changes and everything else is "low". Just to provide some context to that ;-)

**Comment by oddstr13** [comment]
Upstream issue dotnet/corefx#29716 and stale pr dotnet/corefx#29792 tracking this.

**Comment by anthonylavado** [comment]
Shoo bot

**Comment by BaronGreenback** [comment]
Interesting read up on the background of this including some dotnet issues logged in exactly this area.

Just chewing the fat...

From reading up in this, happy eyeballs would have to be manually implemented, as scarey comments like "not implemented in other platforms" are banded around re dns intercepting and retrying on .net platforms.

The good news is that there wouldn't be an additional dns load if we did it the way I propose which is best demonstrated by the following example.

1. HttpClient.SendAsync (for example) is sent passed a url.
2. The host of this url is parsed, resolved, and sorted in the order specified in RFC8305.
3. The url host is replaced by it's Xth ip equivalent (stopping a subsquent dns lookup from occurring)
4. connection attempt made. 
5. Catch exception. If timeout, increase X and repeat until exhausted.

Repeat count and ip depth could be parameterized as well as caching invalid ip addresses, so re-occurrance doesn't cause degredation.

Anyone think of anything i've missed?

**Comment by BaronGreenback** [comment]
Thinking a bit deeper - if a system wide counter was implemented - a pattern could be established and IPv6 requests could be disabled if errors reached a threshold. A literal IPv6 no supported on this network flag.

**Comment by cvium** [comment]
This is now technically possible with .NET 5

New API: https://github.com/dotnet/runtime/issues/41949 https://github.com/dotnet/runtime/pull/42075

Example implementation: https://github.com/ppy/osu-framework/pull/4191/

**Comment by LinAGKar** [comment]
I've just started getting this issue, with Jellyfin 10.8.9, and can't download metadata or update extensions. Even if I disable IPv6 in the Jellyfin settings, Jellyfin still only tries IPv6.

**Comment by cvium** [comment]
Happy Eyeballs was implemented in #8598
